### PR TITLE
CircleCI: Suport macOS

### LINKED ISSUE #5228: Use CircleCI for macOS builds
Thanks to @tresf, we have access to macos builds on CircleCI. We should use it :)

### PR DESCRIPTION
**Pull request by Reflexe**
closes #5228

### CODE REVIEW

**Comment by Reflexe** [comment]
After that PR, we need to drop Travis from the commit status.

File: .circleci/config.yml [hunk 1]
```diff
@@ -34,7 +46,7 @@ shared:
   # Commmon environment variables
   common_environment: &common_environment
     QT5: True
-    CMAKE_OPTS: -DUSE_WERROR=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DUSE_CCACHE=ON
+    CMAKE_OPTS: -DCMAKE_BUILD_TYPE=RelWithDebInfo -DUSE_CCACHE=ON
```
Comment (PhysSong) [inline .circleci/config.yml]: I think removing `-DUSE_WERROR=ON` from Linux/Windows builds is not good. Can we leave this as-is and override it on the Mac section?

File: .circleci/config.yml [hunk 2]
```diff
@@ -12,6 +12,18 @@ shared:
...
+      keys:
+        - homebrew-{{ arch }}-{{ .Environment.CIRCLE_JOB }}-{{ .Branch }}
+        - homebrew-{{ arch }}-{{ .Environment.CIRCLE_JOB }}
+        - homebrew-{{ arch }}
+  save_homebrew_cache: &save_homebrew_cache
+    save_cache:
+      key: homebrew-{{ arch }}-{{ .Environment.CIRCLE_JOB }}-{{ .Branch }}-{{ .BuildNum }}
+      paths:
+        - ~/Library/Caches/Homebrew
+        - /usr/local/Homebrew
```
Comment (PhysSong) [inline .circleci/config.yml]: Caching `~/Library/Caches/Homebrew` sounds good, but does `/usr/local/Homebrew` ever exist?

**Comment by Reflexe** [comment]
This is where homebrew actually downloads its metadata data to. That
reduces setting up by about 5m.

**Comment by tresf** [comment]
@reflexe, the DMG appears to be in working order.

Screen Shot 2019-10-22 at 1 45 32 PM
Screen Shot 2019-10-22 at 1 46 52 PM

File: .circleci/config.yml (hunk 2 above)
Comment (Reflexe) [reply .circleci/config.yml]: This is where homebrew actually downloads its metadata data to. That
reduces setting up by about 5m.

File: .circleci/config.yml (hunk 1 above)
Comment (Reflexe) [reply .circleci/config.yml]: You're right.

File: CMakeLists.txt [hunk 3]
```diff
@@ -491,7 +491,7 @@ FIND_PACKAGE(Samplerate 0.1.8 MODULE REQUIRED)
 
 # set compiler flags
 IF(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
-	SET(WERROR_FLAGS "-Wall -Werror=unused-function -Wno-sign-compare -Wno-strict-overflow")
+	SET(WERROR_FLAGS "-Wall -Werror=unused-function -Wno-sign-compare -Wno-strict-overflow -Wdeprecated-register")
```
Comment (PhysSong) [inline CMakeLists.txt]: Clang emits a ton of errors than this one. Also, did you mean `-Wno-deprecated-register`?

Comment (PhysSong) [reply CMakeLists.txt]: Copying what I said on Discord:
> @Reflexe I'd rather add a `-DUSE_WERROR=OFF` to the macOS build.

File: CMakeLists.txt [hunk 4]
```diff
@@ -493,7 +493,9 @@ FIND_PACKAGE(Samplerate 0.1.8 MODULE REQUIRED)
 IF(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
 	SET(WERROR_FLAGS "-Wall -Werror=unused-function -Wno-sign-compare -Wno-strict-overflow")
 	OPTION(USE_WERROR "Add -werror to the build flags. Stops the build on warnings" OFF)
-	IF(${USE_WERROR})
+
+	# Don't use -Werror with clang since it breaks a lot of stuff.
+	IF(${USE_WERROR} AND NOT CMAKE_CXX_COMPILER_ID EQUAL "Clang")
```
Comment (PhysSong) [inline CMakeLists.txt]: To handle AppleClang:
```suggestion
	IF(${USE_WERROR} AND NOT CMAKE_CXX_COMPILER_ID MATCHES "Clang")
```

**Comment by Reflexe** [comment]
Is there any issue with this PR? can I merge it? 😕

**Comment by PhysSong** [comment]
I can't find any serious issues with this PR. @tresf Could you look at the changes so we can double-check?

**Comment by tresf** [comment]
I'm confused, what part of this build requires CMake changes?  Can't you just set `-DUSE_WERRO=False` on Mac builds?

**Comment by PhysSong** [comment]
> Can't you just set `-DUSE_WERROR=False` on Mac builds?

Do you mean adding the flag after `$CMAKE_OPTS`?

**Comment by tresf** [comment]
I mean setting it in Circle, not CMakeLists.txt

**Comment by Reflexe** [comment]
Our werror options will cause clang to fail. Currently, in circleci, CMAKE_OPTS is a shared for all jobs and specify werror. Instead of changing it, I have decided to fix the root of the problem.

**Comment by PhysSong** [comment]
@Reflexe I think what @tresf mean is:
```diff
-            cmake .. -DCMAKE_INSTALL_PREFIX="../target" -DCMAKE_PREFIX_PATH="$(brew --prefix qt5)" $CMAKE_OPTS
+            cmake .. -DCMAKE_INSTALL_PREFIX="../target" -DCMAKE_PREFIX_PATH="$(brew --prefix qt5)" $CMAKE_OPTS -DUSE_WERROR=OFF
```

**Comment by Reflexe** [comment]
I think that repeating the same argument is a bit strange. Also, we have a problem with clang and werror.

**Comment by tresf** [comment]
> I think that repeating the same argument is a bit strange.

Stranger than changing the CMakeLists (which affects all build environments) for the sole purpose of a CI configuration?

I don't much care as ideally, Clang will some day build with WERROR (@jasp00 actually made it happen, but we got into some disagreements about how to handle upstream build issues).

Anyway, I prefer to handle this environmentally, but won't fight it.

**Comment by Reflexe** [comment]
> > I think that repeating the same argument is a bit strange.
> 
> Stranger than changing the CMakeLists (which affects all build environments) for the sole purpose of a CI configuration?
> 
> I don't much care as ideally, Clang will some day build with WERROR (@jasp00 actually made it happen, but we got into some disagreements about how to handle upstream build issues).
> 
> Anyway, I prefer to handle this environmentally, but won't fight it.

You are right :) I will change it tomorrow.

**Comment by lukas-w** [comment]
What's the benefit of running macOS tests on CircleCI instead of Travis?

**Comment by Reflexe** [comment]
> What's the benefit of running macOS tests on CircleCI instead of Travis?

Reliability and speed. Travis is just hell to work with.

**Comment by tresf** [comment]
> Travis is just hell to work with.

This seems to be modern consensus, but Travis has been a good platform for us for years.

The way I see it is, Circle-CI offers consistency as its primary factor.  For example, the build artifacts would be the influencing factor.  Speed and reliability seem like empty promises.  Each CI will suffer it's own issues on both speed and reliability, both of which fluctuate and are out of our control.

**Comment by tresf** [comment]
> Travis's macOS builds are very inconsistent.

I agree that they do fail to start more often than Linux, but once they start, the usually succeed.  I believe recent failures were due to running an old environment, building qt5 from scratch, because homebrew deliberately breaks older OSs (quite annoying) but would exist on any CI running a dated macOS version.

**Pull request merged**